expr/interval_compound_literal.go

```diff
@@ -0,0 +1,246 @@
+func (m IntervalCompoundLiteral) WithSubSecond(subSeconds int64, precision types.TimePrecision) IntervalCompoundLiteral {
+	return IntervalCompoundLiteral{
+		years:       m.years,
+		months:      m.months,
+		days:        m.days,
+		seconds:     m.seconds,
+		subSeconds:  subSeconds,
+		precision:   precision,
+		nullability: m.nullability,
+	}
+}
+
+func (m IntervalCompoundLiteral) WithNullability(nullability types.Nullability) IntervalCompoundLiteral {
+	return IntervalCompoundLiteral{
+		years:       m.years,
+		months:      m.months,
+		days:        m.days,
+		seconds:     m.seconds,
+		subSeconds:  m.subSeconds,
+		precision:   m.precision,
+		nullability: nullability,
+	}
+}
+
+func (m IntervalCompoundLiteral) getType() types.Type {
+	return types.NewIntervalCompoundType().WithPrecision(m.precision).WithNullability(m.nullability)
+}
+
+func (m IntervalCompoundLiteral) ToProtoLiteral() *proto.Expression_Literal {
+	t := m.getType()
+	intrCompPB := &proto.Expression_Literal_IntervalCompound{}
+
+	if m.years != 0 || m.months != 0 {
+		yearToMonthProto := &proto.Expression_Literal_IntervalYearToMonth{
+			Years:  m.years,
+			Months: m.months,
+		}
+		intrCompPB.IntervalYearToMonth = yearToMonthProto
+	}
+
+	if m.days != 0 || m.seconds != 0 || m.subSeconds != 0 {
+		dayToSecondProto := &proto.Expression_Literal_IntervalDayToSecond{
+			Days:          m.days,
+			Seconds:       m.seconds,
+			PrecisionMode: &proto.Expression_Literal_IntervalDayToSecond_Precision{Precision: m.precision.ToProtoVal()},
+			Subseconds:    m.subSeconds,
+		}
+		intrCompPB.IntervalDayToSecond = dayToSecondProto
+	}
+
+	return &proto.Expression_Literal{
+		LiteralType:            &proto.Expression_Literal_IntervalCompound_{IntervalCompound: intrCompPB},
+		Nullable:               t.GetNullability() == types.NullabilityNullable,
+		TypeVariationReference: t.GetTypeVariationReference(),
+	}
+}
+
+func (m IntervalCompoundLiteral) ToProto() *proto.Expression {
+	return &proto.Expression{RexType: &proto.Expression_Literal_{
+		Literal: m.ToProtoLiteral(),
+	}}
+}
+
+func intervalCompoundLiteralFromProto(l *proto.Expression_Literal) Literal {
+	icLiteral := IntervalCompoundLiteral{}.WithNullability(getNullability(l.Nullable))
+	yearToMonth := l.GetIntervalCompound().GetIntervalYearToMonth()
+	if yearToMonth != nil {
+		icLiteral = icLiteral.WithYears(yearToMonth.Years).WithMonths(yearToMonth.Months)
+	}
+	dayToSecond := l.GetIntervalCompound().GetIntervalDayToSecond()
+	if dayToSecond == nil {
+		// no day to second part
+		return icLiteral
+	}
+	err := validateIntervalDayToSecondProto(dayToSecond)
+	if err != nil {
+		return nil
+	}
+
+	// get subSecond/precision value from proto. To get value it takes care of deprecated microseconds
+	precision, subSeconds, err := intervalCompoundPrecisionSubSecondsFromProto(dayToSecond)
+	if err != nil {
+		return nil
+	}
+	return icLiteral.WithDays(dayToSecond.Days).WithSeconds(dayToSecond.Seconds).WithSubSecond(subSeconds, precision)
```

Comment:
we should avoid the numerous copies this will make and just construct the compound literal directly here instead. As written, this is going to create at least 4 copies of this struct which seems kinda silly

Reply:
fixed it